src/interrupts.rs

```diff
@@ -0,0 +1,192 @@
+/*!
+  # Interrupt Management
+  Interrupts can be enabled, disabled and cleared.
+
+  ## Example
+  ```rust
+    enable_interrupt(TimerCh0);
+
+    // ...
+
+    #[no_mangle]
+    fn TimerCh0() {
+        // ..
+        clear_interrupt(TimerCh0);
+    }
+  ```
+
+  ## The following functions can be implemented as interrupt handlers
+  ```rust
+    fn Gpio();
+    fn TimerCh0();
+    fn TimerCh1();
+  ```
+*/
+
+use riscv::register::mcause;
+
+extern "C" {
+    fn Gpio();
+    fn TimerCh0();
+    fn TimerCh1();
+}
+
+// see components\bl602\bl602_std\bl602_std\RISCV\Core\Include\clic.h
+// see components\hal_drv\bl602_hal\bl_irq.c
+const IRQ_NUM_BASE: u32 = 16;
+const CLIC_HART0_ADDR: u32 = 0x02800000;
+const CLIC_INTIE: u32 = 0x400;
+const CLIC_INTIP: u32 = 0x000;
+const IRQN_LAST: u32 = IRQ_NUM_BASE + 63;
+
+const GPIO_IRQ: u32 = IRQ_NUM_BASE + 44;
+const TIMER_CH0_IRQ: u32 = IRQ_NUM_BASE + 36;
+const TIMER_CH1_IRQ: u32 = IRQ_NUM_BASE + 37;
+
+#[doc(hidden)]
+#[no_mangle]
+pub fn _setup_interrupts() {
+    extern "C" {
+        pub fn _start_trap_hal();
+    }
+
+    let new_mtvec = _start_trap_hal as usize;
+    unsafe {
+        riscv::interrupt::disable();
+        riscv::register::mtvec::write(new_mtvec | 2, riscv::register::mtvec::TrapMode::Direct);
+    }
+
+    // disable all interrupts
+    for i in (0..IRQN_LAST).step_by(4) {
```

Comment:
This way of zeroing out the regs doesn't feel very idiomatic.
I'm not sure of the optimal way of doing this safely, but something using slices would certainly be easier to read.
eg:
`
let e = unsafe {core::slice::from_raw_parts_mut((CLIC_HART0_ADDR + CLIC_INTIE) as *mut u32, 16)};
let p = unsafe {core::slice::from_raw_parts_mut((CLIC_HART0_ADDR + CLIC_INTIP) as *mut u32, 16)};
e.fill(0);
p.fill(0);
`